geomstats/geometry/positive_reals.py

```diff
@@ -0,0 +1,318 @@
+    10.1007/978-3-030-26980-7_27. hal-02875403
+    https://hal.archives-ouvertes.fr/hal-02875403/document
+.. [Cabanes_RADAR_2019] Yann Cabanes, Frédéric Barbaresco, Marc Arnaudon,
+    Jérémie Bigot. Non-Supervised High Resolution Doppler Machine Learning
+    for Pathological Radar Clutter. RADAR 2019, Sep 2020, Toulon, France.
+    10.1109/RADAR41533.2019.171295. hal-02875415
+    https://hal.archives-ouvertes.fr/hal-02875415/document
+.. [JV2016] B. Jeuris and R. Vandebril. The Kahler mean of Block-Toeplitz
+    matrices with Toeplitz structured blocks, 2016.
+    https://epubs.siam.org/doi/pdf/10.1137/15M102112X
+"""
+
+import geomstats.backend as gs
+from geomstats.geometry.base import OpenSet
+from geomstats.geometry.euclidean import Euclidean
+from geomstats.geometry.riemannian_metric import RiemannianMetric
+
+
+class PositiveReals(OpenSet):
+    """Class for the manifold of positive reals.
+
+    The real positive axis endowed with the Information geometry metric.
+
+    Parameters
+    ----------
+    scale : float
+        Scale of the positive reals metric.
+        Optional, default: 1.
+    """
+
+    def __init__(self, scale=1.0, **kwargs):
+        super().__init__(
+            dim=1,
+            embedding_space=Euclidean(1),
+            metric=PositiveRealsMetric(scale=scale),
+            **kwargs
+        )
+        self.scale = scale
+
+    @staticmethod
+    def belongs(point, atol=gs.atol):
+        """Check if a point is a positive real.
+
+        Evaluate if a point belongs to the positive real axis,
+        i.e. evaluate if its norm is lower than one.
+
+        Parameters
+        ----------
+        point : array-like, shape=[..., 1]
+            Points to be checked.
+        atol : float
+            Tolerance.
+            Optional, default: backend atol.
+
+        Returns
+        -------
+        belongs : array-like, shape=[...,]
+            Boolean denoting if point is a positive real.
+        """
+        is_scalar = gs.ndim(point) == 1 or (gs.ndim(point) == 2 and point.shape[1] == 1)
+        if not is_scalar:
+            return gs.zeros([point.shape[0]], dtype=bool)
+        point = gs.reshape(point, (-1,))
+        is_real = gs.imag(point) == 0
+        point = gs.real(point)
+        is_positive = point > 0
+        return gs.logical_and(is_positive, is_real)
+
+    @staticmethod
+    def projection(point, atol=gs.atol):
+        """Project a point on the positive reals.
+
+        Parameters
+        ----------
+        point : array-like, shape=[..., 1]
+            Point in ambient manifold.
+        atol : float
+            Tolerance.
+            Optional, default: backend atol.
+
+        Returns
+        -------
+        projected : array-like, shape=[..., 1]
+            Projected point.
+        """
+        return gs.where(point <= 0, atol, point)
+
+    @staticmethod
+    def random_point(n_samples=1, bound=1.0, atol=gs.atol):
```

Comment:
if we allow `float` above, then here it should retrieve a float for the case `n_samples=1`.

Reply:
Returning a float creates a test error:

`=========================================================================================================== ERRORS ============================================================================================================
________________________________________________________________________________ ERROR collecting tests/tests_geomstats/test_positive_reals.py ________________________________________________________________________________
tests/tests_geomstats/test_positive_reals.py:12: in <module>
    class TestPositiveReals(OpenSetTestCase, metaclass=Parametrizer):
tests/conftest.py:220: in __new__
    test_data = _get_test_data(
tests/conftest.py:305: in _get_test_data
    test_data_ = getattr(testing_data, data_fn_str)()
tests/data_generation.py:266: in random_tangent_vec_is_tangent_test_data
    base_point = space.random_point()
geomstats/geometry/positive_reals.py:140: in random_point
    return (bound - atol) * gs.random.rand() + atol
geomstats/_backend/_dtype_utils.py:249: in _wrapped
    if out.dtype != dtype:
E   AttributeError: 'float' object has no attribute 'dtype'
------------------------------------------------------------------------------------------------------- Captured stderr -------------------------------------------------------------------------------------------------------
INFO: Using numpy backend
=================================================================================================== short test summary info ===================================================================================================
ERROR tests/tests_geomstats/test_positive_reals.py - AttributeError: 'float' object has no attribute 'dtype'
`

Reply:
So I did not allow to return a float. Done.